📌 [Week4] 서버 통신을 통해 회원가입/로그인/내 정보 구현하기

hyeminililo opened the pull request:
### 🫛 작업한 내용
  - 로그인/회원가입 응답 모델 정리
    로그인, 회원가입 API의 JSON 구조에 맞춰` DTO(ResponseLoginDto, ResponseUserDto) `객체를 재구성하고, 내부 User 객체도 재구성했습니다. 
  - ViewModel 상태 및 흐름 개선
	  3주차 과제 리팩토링을 하면서 `ViewModel`에 대해 전보다 조금 더 이해한 것 같습니다. 그래서 이와 관련해 코드리뷰를 기반으로 리팩토링 진행했습니다. 

  ### 🫛 PR 포인트
  아래 궁금하고 어려웠던 점 부분을 위주로 봐주시면 감사하겠습니다 🙇🏻‍♀️

  ### 📸 스크린샷

<details>
  <summary>영상</summary>
     
https://github.com/user-attachments/assets/7de79e70-ee5b-42d2-bdaa-26c107b6f309

</details>


  ### 🤔 궁금하고 어려웠던 점

  1. 오류 DTO 작성
	서버 에러 응답을 통일해서 DTO로 받으면 좋겠다고 판단해 구현했지만, 어떻게 활용해야 하는지 어려웠습니다.

  2.  `private val json = Json { explicitNulls = false }` 사용 
	 API를 연결하면서 연결을 확인하는데 앱이 중단되는 상황이 계속 있었습니다. 그래서 이후, 구글링을 해보면서 이런 방법을 하면 연결이 된다고 해서 사용했는데, 사용해도 괜찮은지 궁금합니다 😭

 3. Navigator와 viewmodel의 역할분담 
	 이번 과제에서도 가장 어려웠던 부분입니다. `ViewModel`을 `상태`를 관리하는 주체로 두고,  네트워크 응답을 받은 뒤 상태를 업데이트하는 흐름 자체로 코드를 구현했습니다. 하지만 아래 4개가 조금 어려웠습니다. 
	
	-   API 응답이 성공관련 로직은 어디에 짜야하는지 ?
	-   데이터 응답값을 잘 받아왔는지 확인하는 로직은 어디에 짜야하는지 ?
	-   로그인 성공 ->  개인 정보 조회 -> 화면 이동, 이 흐름을 어떻게 나누는지?
	- 
 이 역할 분리 기준이 아직 명확하지 않아서,  관련해 코드 리뷰를 받고 싶습니다..

4. User와 RequestUser가 만약 같다면 하나로 통일 해서 사용해도 되는지 
현재 프로젝트에는 앱 내부에서 사용하는 `User`, 서버 요청/응답에 사용하는 `RequestUserDto`가 2개가 있습니다.
이렇게 두 종류의 모델이 있습니다. 만약 requestUser와 앱 내부 User의 구조가 완전히 같다면  두 개로 나누어도 괜찮은지 궁금했습니다.  중복 코드가 많아지면 유지보수가 어려워지긴 하지만 제가 생각했을 때는 서버와 통신하는 DTO와 앱 내부에서 사용하는 객체는 다른 객체를 사용하는 것이 더 객체지향적이라 생각했기 때문입니다..! 

  ### 📮 관련 이슈
  #9

hyeminililo committed:
#9 chore: 인터넷 권한 설정 및 통신 관련 의존성 추가

hyeminililo committed:
#9 feat: 회원가입 응답, 요청 DTO 구현

hyeminililo committed:
#9 feat: 로그인 응답, 요청 DTO 구현

hyeminililo committed:
#9 rename: responseSignupDto -> responseUserDto 이름 변경

개인 정보 조회와 응답값이 동일해, 기존 회원가입 응답 DTO 이름 변경

hyeminililo committed:
#9 feat: 응답시, 에러 상황  DTO 구현

hyeminililo committed:
#9 feat: AuthService, UserService 구현

HTTP Method를 정의해둔 auth, user 인터페이스 작성

hyeminililo committed:
#9 feat: API Factory 구현

Retrofit 서비스 객체를 사용하기위한 인터페이스 객체 구현

hyeminililo committed:
#9 refactor: User 모델 변경

git add .

hyeminililo committed:
#9 refactor: User 모델 변경

nickname, birthday -> name, email, age로 변경

hyeminililo committed:
#9 refactor: 변경된 모델에 맞게 화면 변경

hyeminililo committed:
#9 feat: User 레포지토리 구현

데이터 보내는 레포지토리 구현

hyeminililo committed:
#9 feat: Auth 레포지토리 구현

데이터 보내는 레포지토리 구현

hyeminililo committed:
#9 build: BASE_URL 변수 수정

hyeminililo committed:
#9 fix: api 주소 수정

hyeminililo committed:
#9 feat: User 모델 서버 값에 맞게 수정

-  기존 nickname, birthday 삭제
-  새로운 username, age, email 생성

hyeminililo committed:
#9 feat: API 연동 로직 ViewModel에 추가

 로그인, 회원가입, 유저 조회등 API 연동 로직을 UserViewModel.kt에 추가

hyeminililo committed:
#9 feat: 서버 모델에 맞춰 변경

hyeminililo committed:
#9 feat: ResponseDto 변경

hyeminililo committed:
#9 chore: http 통신 허용을 위한 xml 추가

hyeminililo committed:
#9 delete: 안 쓰는 repository 삭제

hyeminililo committed:
#9 refactor: 불필요한 공백 및 임포트 정리

oilbeaneda commented on app/src/main/java/com/sopt/dive/ui/components/CommonButton.kt:
```diff
@@ -11,12 +11,12 @@ import com.sopt.dive.ui.theme.MainPinkBackground
-        contentColor = Color.White
-    )
-        ,
+            containerColor = MainPinkBackground,
```
컬러 시스템 바로 세팅하신거 넘 멋져요!!! 저도 본받아야겠두요

oilbeaneda reviewed (commented):
늘 열심히 하시는 혜민님 넘 고생많으셨습니다!! 네트워크 통신 쪽 이해가 맣아ㅏ아아않이 어려워서.. 월화중으로 더 공부해서 질문 남겨놓겠습니다!!

oilbeaneda commented on app/src/main/java/com/sopt/dive/viewmodel/UserViewModel.kt:
```diff
@@ -1,41 +1,135 @@
+            }
+        })
     }
+    private fun ResponseUserDto.toUser(): User {
```
바로위 fetchUser에서 데이터를 받아오는 것이아닌 ResponseUserDto에서 값을 전달 받는 것인지 궁금합니다! 그렇다면 한번에가아닌 따로 해야하는 이유가 있는지도 궁금합니다🤩

hyeminililo replied:
제가 받은 DTO를  앱 내부에서 쓰기 위해서 사용했어요 !!

oilbeaneda reviewed (commented):
늘 열심히 하시는 혜민님 넘 고생많으셨습니다!! 컬러시스템 바로 적용하신 것두 넘 멋져요 ㅎㅎ 저도 본받아야겠두요🤩
그리고 네트워크 통신 쪽 이해가 맣아ㅏ아아않이 어려워서.. 제가 월화중으로 더 공부해서 질문 남겨놓겠습니다!!

sonms commented on app/src/main/java/com/sopt/dive/data/ApiFactory.kt:
```diff
@@ -0,0 +1,49 @@
+    val authService: AuthService by lazy {
+        ApiFactory.create<AuthService>()
+    }
+}
```
ApiFactory의 역할과 servicepool의 역할이 같을까요~? 저는 분리하면 좋을 것 같습니다!

hyeminililo replied:
넵 !

sonms commented on app/src/main/java/com/sopt/dive/data/api/AuthService.kt:
```diff
@@ -0,0 +1,14 @@
+    @POST("api/v1/auth/login")
+    fun login(
+        @Body request: RequestLoginDto
+    ): Call<ResponseLoginDto>
```
Call말고 Response도 존재하니 확인해보시면 좋을 것 같습니다~

sonms replied:
그리고 현재 Callback 방식을 사용하고 계신데 코드가 길어지고 에러 처리와 성공 처리 분리가 필요해요
그래서 코루틴 방식으로 구현해서 블로킹 없이 백그라운드에서 실행하고 결과도 바로 반환 받고 실패 처리도 쉽게 구현할 수 있도록 구현해보셨으면 좋겠습니다!

hyeminililo replied:
6주차 과제로 적용해보겠습니다 !

hyeminililo replied:
찾아보니 Call은 조금 오래된 방식이군요 ,, Response를 이용해보도록 하겠습니다 :)

sonms commented on app/src/main/java/com/sopt/dive/data/dto/ResponseUserDto.kt:
```diff
@@ -0,0 +1,42 @@
+    val data: ResponseUserDto?
+)
+
+enum class Status { ACTIVE }
```
여기서 만들기 보다는 viewmodel에서 결과를 받으면서 파싱해보면 어떨까요?

hyeminililo replied:
이번 합세를 하면서 궁금한 점이 생겼습니다 .. Mapper 라는 것을 활용해봤는데요 ,, Mapper와 viewmodel 중 어떤 것이 더 적절할까요오 ,,?

sonms commented on app/src/main/java/com/sopt/dive/ui/screens/LoginScreen.kt:
```diff
@@ -51,27 +52,30 @@ fun LoginScreen(
-                } else {
-                    Toast.makeText(context, "로그인 실패", Toast.LENGTH_SHORT).show()
-                }
+                val request = RequestLoginDto(idText, pwText)
```
Dto란 무엇일까요? presentation에서 dto 객체를 직접생성해서 사용해도 괜찮을까요~? 좀 더 고민해보셨으면 좋겠습니다~

hyeminililo replied:
제가 질문 드린걸 다시 생각해보니 Dto는 Presentation에 영향을 주면 안 될 것같네요 ! 분리를 해야할 것 같다는 생각입니다 ,,!

sonms reviewed (commented):
어려운데도 적용해보고 고민해보시는 모습이 정말 좋습니다!! 혜민님도 이번에 합동 세미나에서 더욱 성장하시길 바라겠습니다~! 👍 조금 더 자세히 읽고 리뷰 더 작성하도록 하겠습니다!

sohee6989 commented on app/src/main/java/com/sopt/dive/ui/components/CommonButton.kt:
```diff
@@ -11,12 +11,12 @@ import com.sopt.dive.ui.theme.MainPinkBackground
 fun CommonButton(textMessage: String, onClick: () -> Unit, modifier: Modifier = Modifier) {
-    Button(onClick = onClick, modifier.fillMaxWidth(),
+    Button(
+        onClick = onClick, modifier.fillMaxWidth(),
```
일관되게 modifier도 파라미터 명시해주면 좋을 것 같습니다!

sohee6989 commented on app/src/main/java/com/sopt/dive/data/dto/ResponseUserDto.kt:
```diff
@@ -0,0 +1,42 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class ResponseUserBodyDto(
```
이 부분 다른 부분들에서도 공통적으로 사용되고 있는 부분인데 공통 response로 분리해주는거 어떠세요??

hyeminililo replied:
넵 분리해보겠습니다. !

sohee6989 commented on app/src/main/java/com/sopt/dive/model/UserProfile.kt:
```diff
@@ -4,5 +4,5 @@ import android.os.Parcelable
-data class HomeProfileInfo(val userName: String, val title: String, val content: String) : Parcelable
-
+data class HomeProfileInfo(val userName: String, val title: String, val content: String) :
+    Parcelable
```
이 Parcelize 어노테이션의 역할이 궁금합니다!

sohee6989 commented on app/src/main/java/com/sopt/dive/data/dto/ResponseLoginDto.kt:
```diff
@@ -0,0 +1,28 @@
+@Serializable
+data class LoginDataDto(
+    @SerialName("userId")
+    val userId: Int,
```
api 명세서에 userId 타입이 Long 아니였나요??

hyeminililo replied:
하하 그러쵸 ,, 제가  Int로 생각했네요 스정하겠습니다 😭

sohee6989 reviewed (commented):
수고하셨습니다!! 코드 잘 봤습니다!
아직 코드를 다 못 봐서 추가로 좀 더 남기러 곧 오겠습니다ㅎ.ㅎ

sonms commented on app/src/main/java/com/sopt/dive/data/api/UserService.kt:
```diff
@@ -0,0 +1,21 @@
+
+import com.sopt.dive.data.dto.RequestSignupDto
+import com.sopt.dive.data.dto.ResponseUserBodyDto
+import retrofit2.Call // 요 레트로핏 맞겠지 ?
```
마자요~

sonms commented on app/src/main/java/com/sopt/dive/data/dto/ResponseErrorDto.kt:
```diff
@@ -0,0 +1,49 @@
+    val success: Boolean = false,
+
+    @SerialName("code")
+    val code: String,
+
+    @SerialName("message")
+    val message: String,
+
+    @SerialName("data")
+    val data: ErrorDataDto?
+
+)
+
+@Serializable
+data class ErrorDataDto(
+
+    @SerialName("code")
+    val code: String,
+
+    @SerialName("message")
+    val message: String,
+
+    @SerialName("errors")
+    val errors: List<FieldErrorDto>?
+
+)
+
+@Serializable
+data class FieldErrorDto(
+
+    @SerialName("field")
+    val field: String,
+
+    @SerialName("value")
+    val value: String,
+
+    @SerialName("reason")
+    val reason: String
+
+)
```
그렇기에 dto로 응답을 받을 때 T라는 제네릭을 통해 받습니다!
보통 서버에서 오는 응답은 다음과 같은 구조를 가져요

{
  "code": 200,
  "message": "success",
  "data": {...}
}


여기서 data 부분은 요청마다 다른 타입일 수 있는데 이를 위해 T를 사용하면
T에 어떤 타입이 들어올지는 호출 시점에서 정하기 때문에 여러 타입을 안전하게 처리할 수 있어요

hyeminililo replied:
넵 적용해보겠습니다 !

sonms commented on app/src/main/res/xml/network_security_config.xml:
```diff
@@ -0,0 +1,6 @@
+    <domain-config cleartextTrafficPermitted="true">
+        <domain includeSubdomains="true">15.164.129.239</domain>
+    </domain-config>
+</network-security-config>
```
이 코드 없이 manifest에` android:usesCleartextTraffic="true"` 만으로도 괜찮을 것 같아요~

hyeminililo replied:
오 꿀팁 감사합니다 !

sonms commented on app/src/main/java/com/sopt/dive/util/Navigator.kt:
```diff
@@ -99,29 +101,48 @@ fun Navigator(navController: NavHostController = rememberNavController()) {
             startDestination = Route.Login.path
         ) {
             composable(Route.Login.path) {
+                val loginSuccess by userViewModel.loginSuccess.collectAsState()
```
collectAsState와 collectAsStateWithLifecycle은 무슨 차이가 있을까요? 한 번 확인해보셨으면 좋겠습니다 ! 사실 이름에 힌트가 있어요

hyeminililo replied:
일단 찾아보기 전에는 `collectAsStateWithLifecycle`이름만 들었을 때는 LifeCycle을 관리할 수 있다는 것 같네요 ,,,! 

그리고 찾아보니 해당 LifeCycle 상태에 맞춰서 흐름을 관찰하고 데이터를 받는 것 같아요 !

sonms reviewed (commented):
> 오류 DTO 작성
서버 에러 응답을 통일해서 DTO로 받으면 좋겠다고 판단해 구현했지만, 어떻게 활용해야 하는지 어려웠습니다.

활용하는 방식으로는 제네릭 T 가 있어요 지금 코드보면 모든 응답 구조를 Error와 성공했을 때를 만드신 것 같은데 제네릭을 통해 
```kotlin
class Box<T>(val value: T)

// 사용 시점에 타입을 정함
val intBox = Box(123)       // T = Int
val stringBox = Box("Hi")   // T = String
```
이런식으로 사용할 시점 즉, 데이터를 불러온 시점에서 타입을 설정할 수 있도록 하면 활용성이 높아집니다!

> private val json = Json { explicitNulls = false } 사용
API를 연결하면서 연결을 확인하는데 앱이 중단되는 상황이 계속 있었습니다. 그래서 이후, 구글링을 해보면서 이런 방법을 하면 연결이 된다고 해서 사용했는데, 사용해도 괜찮은지 궁금합니다 😭

이 설정 같은 경우 JSON 직렬화 시 null 필드를 보내느냐 안 보내느냐를 결정하는 설정인거 같은데 우선 어떤 오류가 발생하는지도 궁금하네요 ㅎㅎ 우선 이 코드만 보고 말씀드리자면 사용하는 것도 괜찮습니다! 
다만 서버와 통신하는 부분이기 때문에 서버 파트와의 대화 후에 괜찮겠다 싶으시면 사용하시는 것을 추천드려요~ 혹시나 서버에 문제가 생길 수도 있으니까요 :)

> Navigator와 viewmodel의 역할분담
이번 과제에서도 가장 어려웠던 부분입니다. ViewModel을 상태를 관리하는 주체로 두고, 네트워크 응답을 받은 뒤 상태를 업데이트하는 흐름 자체로 코드를 구현했습니다. 하지만 아래 4개가 조금 어려웠습니다.
API 응답이 성공관련 로직은 어디에 짜야하는지 ?
데이터 응답값을 잘 받아왔는지 확인하는 로직은 어디에 짜야하는지 ?
로그인 성공 -> 개인 정보 조회 -> 화면 이동, 이 흐름을 어떻게 나누는지?
이 역할 분리 기준이 아직 명확하지 않아서, 관련해 코드 리뷰를 받고 싶습니다..

우선 ViewModel은 SSOT(단일 진실 공급원)으로 화면에 보여지는 모든 데이터(텍스트, 색상, 로딩바 여부 등)는 오직 ViewModel만이 가지고 있고 관리해야 한다는 원칙이에요 
동기화 문제, 생명주기, 사이드 이펙트 등과 분리해서 UI를 더 테스트하기 쉽게 만들어줘요

여기서 네비게이션으로 넘기는 값도 포함이에요 이러한 값은 ViewModel에서 관리 가능하며 필요 시 SavedStateHandle을 사용해 화면 간 상태 전달 및 복원을 안전하게 처리할 수 잇어요

그렇기에 1. API 응답이 성공관련 로직은 어디에 짜야하는지 ?는 ViewModel에서 작성할 수 있어요 , 통신과 관련한 비즈니스 로직일 뿐더러 데이터 통신이 받은 데이터를 단일 진실 공급원에서 관리하기 위해서에요 

2. 데이터 응답값을 잘 받아왔는지 확인하는 로직은 어디에 짜야하는지 ? 어디에서 처리하냐에 따라 그 목적이 달라져요
 data 레이어 즉, 추후에 나올 datasource나 repository 단에서 처리하는 방법으로 **유효한 데이터**인지를 확인하는 방식이 있어요 (응답이 200인지, json 구조가 올바른지, null 체크 등등) ViewModel이 네트워크나 데이터 구조에 종속되지 않을 수 있겠지만 그러면 값 처리가 어려워지기 때문에 저는 유효한 값인지, 응답이 성공했는지 등 만 확인 시 사용해요(가끔)
 
 이제 ViewModel에서 처리하는 방법이 있어요 Result 를 통해 우리는 이제 try-catch문이 아니라 
성공(onSuccess)과 실패(onFailure)를 체이닝 형식 및 하나의 객체로 안전하게 표현이 가능해요! 
여기서 성공 시 UI의 데이터 업데이트 그리고 실패 시 사용자에게 실패 알림 등 다양한 UI 관련 이벤트 처리 등이 가능해요

즉, ViewModel이 ‘무엇을 보여줄지, 어떤 흐름이 필요한지’를 관리한다면 Navigator는 실제 화면 이동만 담당해요!

3. 로그인 성공 -> 개인 정보 조회 -> 화면 이동, 이 흐름을 어떻게 나누는지?
처음 로그인 성공 시 서버에서는 유저 정보 조회에 필요한 id값을 response로 줄거에요 그 id를 datastore 이나 sharedPreference에 저장해두고 개인 정보 조회하는 myPage에 진입 시 그 id 를 불러와서 서버와 통신한 후 응답값을 통해 UI를 표시해요!
이러한 흐름을 거치게 된다면 화면 이동 시에는 큰 불편이 없을 거에요 

지금 코드에는 로그인 부분에 많은 책임이 담겨져 있어 그 데이터를 여기저기 끌어오고 들고가느라 너무 코드가 복잡해 보이는 것 뿐이에요! 조금 더 개선해본다면 혜민님도 흐름에 대해 이해하실 수 있으실거에요!

> User와 RequestUser가 만약 같다면 하나로 통일 해서 사용해도 되는지
현재 프로젝트에는 앱 내부에서 사용하는 User, 서버 요청/응답에 사용하는 RequestUserDto가 2개가 있습니다.
이렇게 두 종류의 모델이 있습니다. 만약 requestUser와 앱 내부 User의 구조가 완전히 같다면 두 개로 나누어도 괜찮은지 궁금했습니다. 중복 코드가 많아지면 유지보수가 어려워지긴 하지만 제가 생각했을 때는 서버와 통신하는 DTO와 앱 내부에서 사용하는 객체는 다른 객체를 사용하는 것이 더 객체지향적이라 생각했기 때문입니다..!

맞아요! Dto는 서버 변경 독립성을 위해 서버 API가 변경되서 Dto를 수정하게 되어도 앱 내부 모델은 영향 받지 않고 명확한 책임과 테스트 용이성을 위해 분리하는게 좋아요


Serializable로 구현된 dto를 ui에서 state로 사용하게 된다면 UI에서는 JSON 직렬화/역직렬화가 필요 없는데 그대로 사용되겠죠? 또한 서버에서 구조가 변경되어 dto를 완전히 갈아엎는다면 ui 도 그에 따라 변경되어야 하겠죠

다만 혜민님의 코드에서는 거의 모든 부분에서 Dto를 사용하고 계신 모습이 보입니다! 😢
혜민님의 생각은 저와 같은 생각이에요 그 부분을 코드로 표현해보실 수 있었으면 좋겠습니다 :)

hyeminililo committed:
refactor: 코드리뷰 기반 리팩토링

hyeminililo committed:
#11 refactor: 코드 리뷰 기반으로 로그인 로직 수정

hyeminililo committed:
#11 refactor: 불필요한 코드 삭제

hyeminililo committed:
#10 #11 mod: Call -> Response 변경

hyeminililo committed:
#10 #11 refactor: 코드 리팩토링

hyeminililo committed:
#10 #11 merge  branch 'dev' into week4
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.
